src/Query.re

```diff
@@ -3,68 +3,74 @@ open Revery.UI.React;
 include S;
 
 module type Query = {
-  type t;
+  type t('responseType) = 'responseType;
 
-  type status =
+  type status('responseType) =
     | Idle
     | Loading
     | Error
-    | Data(t);
+    | Data(t('responseType));
 
-  let use:
+  let useQuery:
     (
       ~variables: Yojson.Basic.t=?,
+      (Yojson.Basic.t => t('responseType), string, 'b),
       unit,
       Hooks.t(
         (
-          Hooks.Reducer.t(status),
-          Hooks.Effect.t(Hooks.Effect.onMount),
+          Hooks.Reducer.t(status('responseType)),
           Hooks.Effect.t(option(Yojson.Basic.t))
         ) =>
-        'a,
-        'b,
+        'c,
+        'd,
       )
     ) =>
-    (status, Hooks.t('a, 'b));
+    (status('responseType), Hooks.t('c, 'd));
 };
 
-module Make = (C: BaseConfig, G: QueryConfig) : (Query with type t = G.t) => {
-  type t = G.t;
+module Make = (C: BaseConfig) : Query => {
   let baseUrl = C.baseUrl;
   let headers = C.headers;
 
-  type status =
+  type t('responseType) = 'responseType;
+
+  type status('responseType) =
     | Idle
     | Loading
     | Error
-    | Data(G.t);
+    | Data(t('responseType));
 
-  type action =
-    | Fetch
+  type action('responseType) =
+    | FetchData
     | Error
-    | Data(G.t);
+    | Data('responseType);
 
-  let reducer = (action, _state): status =>
+  let reducer = (action, _state): status('responseType) =>
     switch (action) {
-    | Fetch => Loading
+    | FetchData => Loading
     | Error => Error
     | Data(obj) => Data(obj)
     };
 
-  let initialState: status = Idle;
+  let initialState: status('responseType) = Idle;
+
+  let useQuery = (~variables: option(Yojson.Basic.t)=?, definition, ()) => {
+    let (parse, graphqlQuery, composeVariables) = definition;
 
-  let use = (~variables: option(Yojson.Basic.t)=?, ()) => {
     let%hook (state, dispatch) = Hooks.reducer(~initialState, reducer);
 
-    let query = () => {
+    let queryAndVariables = () => {
```

Comment:
I made this a pure function outside of the module btw. Seemed cleaner as it won't shadow `variables` and `query` and shortens the body of `useQuery`:

```
let createBody = (~variables=`Assoc([]), ~query) => {
  let payload =
    `Assoc([("query", `String(query)), ("variables", variables)])
    |> Yojson.Basic.to_string;

  (payload, variables);
};
```

Given `~variables` a default argument will mean it needs to be called like this further down:

```
       let (payload, variables) = createBody(~query, ~variables?);
```

@lessp What do you think?

Reply:
Agree! Do you mind adding it yourself?

Reply:
Resolved in: fd8d9b9, thanks!